security/default-backend/ndn-lite-default-ecc-impl.c

```diff
@@ -265,19 +205,20 @@ ndn_lite_default_ecdsa_sign(const uint8_t* input_value, uint32_t input_size,
   ctx->uECC.block_size = NDN_SEC_ECC_SECP256R1_PUBLIC_KEY_SIZE;
   ctx->uECC.result_size = NDN_SEC_ECC_SECP256R1_PRIVATE_KEY_SIZE;
   ctx->uECC.tmp = tmp;
-  ecc_sign_result = uECC_sign_deterministic(abs_key->key_value, input_value, input_size,
-                                            &ctx->uECC, output_value, curve);
+
+  uECC_set_rng(&fake_rng);
```

Comment:
Deterministic signing does not require RNG to be set.

Reply:
Yes it shouldn't have required RNG to be set. What's weird is that if I commented it out, I always got a Segmentation fault: 11 during signing (in unit test). Will try to fix this quickly.

Reply:
This indicates your test cases are still using code paths that require RNG, while hardware RNG is absent.
To effectively test all code paths, there should be separate test suites: one compiled with hardware RNG (on Linux it would be OpenSSL RNG) testing normal signing, the other compiled without hardware RNG testing deterministic signing.

Reply:
Thanks for your suggestion!

I just figured it out why: it was caused by both micro-ecc 'static' slight different  implementation and design flaw regarding security module initialization.

- **'static ' branch is slight different** 
It tries to guess the system you run and implement corresponding [default RNG function](https://github.com/kmackay/micro-ecc/blob/static/uECC.c#L423-L458). For unrecognized platforms, simply return 0. So it directly calls RNG in implementation, even in deterministic signing.

- **security module isn't correctly initialized**
RNG doesn't have default backend, but the `ndn_rng_backend` is still there without assignment. Thus according to [current implementation](https://github.com/named-data-iot/ndn-lite/blob/master/security/ndn-lite-sec-config.c#L43-L46), micro-ecc will be set with a null RNG, which overrides the default setting mentioned above.

In the past it worked fine. That's because `ndn_rng_backend` was the only way to bind platform-specific RNG to NDN-Lite, but now micro-ecc 'static' wants to do the binding also. Therefore the conflict happens here.

Reply:
To protected against this case, you'll need a default RNG backend that always fails.